clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -246,6 +252,10 @@ private void validateRecords(short version, MemoryRecords records) {
             if (entry.magic() != RecordBatch.MAGIC_VALUE_V2)
                 throw new InvalidRecordException("Produce requests with version " + version + " are only allowed to " +
                         "contain record batches with magic version 2");
+            if (version < 7 && entry.compressionType() == CompressionType.ZSTD) {
+                throw new InvalidRecordException("Produce requests with version " + version + " are note allowed to " +
```

Comment:
Can we use `UnsupportedVersionException`? The record is valid, it is just that the produce request version doesn't support it.